### ISSUE TITLE
Relocation unit test failing on local development box

### ISSUE THREAD

**Issue by alalazo** [spack/spack#12923]
Updating to the latest version of `develop` I got relocation unit tests failing on my development machine.

### Steps to reproduce the issue
Run the following command:
```console
$ spack test -k relocate
```

### Error message
<details>

<summary>Click to expand the error message</summary>

```console
============================================================== test session starts ===============================================================
platform linux2 -- Python 2.7.15+, pytest-3.2.5, py-1.4.34, pluggy-0.4.0
rootdir: /home/culpo/PycharmProjects/spack/lib/spack/spack/test, inifile: pytest.ini
collected 1806 items                                                                                                                              

packaging.py ..s
relocate.py FF.
============================================================ short test summary info =============================================================
FAIL relocate.py::test_file_is_relocatable[True]
FAIL relocate.py::test_file_is_relocatable[False]
SKIP [1] packaging.py:362: only works with Mach-o objects

=========================================================== slowest 20 test durations ============================================================
0.07s call     relocate.py::test_file_is_relocatable[True]
0.07s call     relocate.py::test_file_is_relocatable[False]
0.01s setup    relocate.py::test_file_is_relocatable[False]
0.01s call     packaging.py::test_relocate_text
0.00s teardown packaging.py::test_relocate_text
0.00s setup    packaging.py::test_relocate_text
0.00s setup    packaging.py::test_relocate_links
0.00s setup    relocate.py::test_file_is_relocatable[True]
0.00s setup    relocate.py::test_file_is_relocatable_errors
0.00s teardown relocate.py::test_file_is_relocatable[False]
0.00s teardown relocate.py::test_file_is_relocatable[True]
0.00s teardown packaging.py::test_relocate_links
0.00s teardown relocate.py::test_file_is_relocatable_errors
0.00s call     relocate.py::test_file_is_relocatable_errors
0.00s teardown packaging.py::test_relocate_macho
0.00s call     packaging.py::test_relocate_links
0.00s setup    packaging.py::test_relocate_macho
==================================================================== FAILURES ====================================================================
_________________________________________________________ test_file_is_relocatable[True] _________________________________________________________

source_file = local('/tmp/pytest-of-culpo/pytest-43/test_file_is_relocatable_True_0/relocatable.c'), is_relocatable = True

    @pytest.mark.requires_executables(
        '/usr/bin/gcc', 'patchelf', 'strings', 'file'
    )
    def test_file_is_relocatable(source_file, is_relocatable):
        compiler = spack.util.executable.Executable('/usr/bin/gcc')
        executable = str(source_file).replace('.c', '.x')
        compiler_env = {
            'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
        }
        compiler(str(source_file), '-o', executable, env=compiler_env)
    
        assert spack.relocate.is_binary(executable)
>       assert spack.relocate.file_is_relocatable(executable) is is_relocatable

relocate.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../relocate.py:678: in file_is_relocatable
    rpaths = patchelf('--print-rpath', file, output=str).strip()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <exe: ['None']>, args = ('--print-rpath', '/tmp/pytest-of-culpo/pytest-43/test_file_is_relocatable_True_0/relocatable.x'), kwargs = {}
env_arg = None
env = {'BASH_FUNC_spack%%': '() {  if [ -n "${ZSH_VERSION:-}" ]; then\n emulate -L sh;\n fi;\n _sp_flags="";\n while [ ! -z ... 'CLUTTER_IM_MODULE': 'xim', 'COLORTERM': 'truecolor', 'DBUS_SESSION_BUS_ADDRESS': 'unix:path=/run/user/1000/bus', ...}
fail_on_error = True, ignore_errors = (), input = None, output = <type 'str'>, error = None, streamify = <function streamify at 0x7fd09ec49758>
ostream = -1, close_ostream = False

    def __call__(self, *args, **kwargs):
        """Run this executable in a subprocess.
    
            Parameters:
                *args (str): Command-line arguments to the executable to run
    
            Keyword Arguments:
                _dump_env (dict): Dict to be set to the environment actually
                    used (envisaged for testing purposes only)
                env (dict): The environment to run the executable with
                extra_env (dict): Extra items to add to the environment
                    (neither requires nor precludes env)
                fail_on_error (bool): Raise an exception if the subprocess returns
                    an error. Default is True. The return code is available as
                    ``exe.returncode``
                ignore_errors (int or list): A list of error codes to ignore.
                    If these codes are returned, this process will not raise
                    an exception even if ``fail_on_error`` is set to ``True``
                input: Where to read stdin from
                output: Where to send stdout
                error: Where to send stderr
    
            Accepted values for input, output, and error:
    
            * python streams, e.g. open Python file objects, or ``os.devnull``
            * filenames, which will be automatically opened for writing
            * ``str``, as in the Python string type. If you set these to ``str``,
              output and error will be written to pipes and returned as a string.
              If both ``output`` and ``error`` are set to ``str``, then one string
              is returned containing output concatenated with error. Not valid
              for ``input``
    
            By default, the subprocess inherits the parent's file descriptors.
    
            """
        # Environment
        env_arg = kwargs.get('env', None)
        if env_arg is None:
            env = os.environ.copy()
            env.update(self.default_env)
        else:
            env = self.default_env.copy()
            env.update(env_arg)
        env.update(kwargs.get('extra_env', {}))
        if '_dump_env' in kwargs:
            kwargs['_dump_env'].clear()
            kwargs['_dump_env'].update(env)
    
        fail_on_error = kwargs.pop('fail_on_error', True)
        ignore_errors = kwargs.pop('ignore_errors', ())
    
        # If they just want to ignore one error code, make it a tuple.
        if isinstance(ignore_errors, int):
            ignore_errors = (ignore_errors, )
    
        input  = kwargs.pop('input',  None)
        output = kwargs.pop('output', None)
        error  = kwargs.pop('error',  None)
    
        if input is str:
            raise ValueError('Cannot use `str` as input stream.')
    
        def streamify(arg, mode):
            if isinstance(arg, string_types):
                return open(arg, mode), True
            elif arg is str:
                return subprocess.PIPE, False
            else:
                return arg, False
    
        ostream, close_ostream = streamify(output, 'w')
        estream, close_estream = streamify(error,  'w')
        istream, close_istream = streamify(input,  'r')
    
        quoted_args = [arg for arg in args if re.search(r'^"|^\'|"$|\'$', arg)]
        if quoted_args:
            tty.warn(
                "Quotes in command arguments can confuse scripts like"
                " configure.",
                "The following arguments may cause problems when executed:",
                str("\n".join(["    " + arg for arg in quoted_args])),
                "Quotes aren't needed because spack doesn't use a shell.",
                "Consider removing them")
    
        cmd = self.exe + list(args)
    
        cmd_line = "'%s'" % "' '".join(
            map(lambda arg: arg.replace("'", "'\"'\"'"), cmd))
    
        tty.debug(cmd_line)
    
        try:
            proc = subprocess.Popen(
                cmd,
                stdin=istream,
                stderr=estream,
                stdout=ostream,
                env=env)
            out, err = proc.communicate()
    
            result = None
            if output is str or error is str:
                result = ''
                if output is str:
                    result += text_type(out.decode('utf-8'))
                if error is str:
                    result += text_type(err.decode('utf-8'))
    
            rc = self.returncode = proc.returncode
            if fail_on_error and rc != 0 and (rc not in ignore_errors):
                long_msg = cmd_line
                if result:
                    # If the output is not captured in the result, it will have
                    # been stored either in the specified files (e.g. if
                    # 'output' specifies a file) or written to the parent's
                    # stdout/stderr (e.g. if 'output' is not specified)
                    long_msg += '\n' + result
    
                raise ProcessError('Command exited with status %d:' %
                                   proc.returncode, long_msg)
    
            return result
    
        except OSError as e:
            raise ProcessError(
>               '%s: %s' % (self.exe[0], e.strerror), 'Command: ' + cmd_line)
E           ProcessError: None: No such file or directory
E               Command: 'None' '--print-rpath' '/tmp/pytest-of-culpo/pytest-43/test_file_is_relocatable_True_0/relocatable.x'

../util/executable.py:195: ProcessError
________________________________________________________ test_file_is_relocatable[False] _________________________________________________________

source_file = local('/tmp/pytest-of-culpo/pytest-43/test_file_is_relocatable_False0/non_relocatable.c'), is_relocatable = False

    @pytest.mark.requires_executables(
        '/usr/bin/gcc', 'patchelf', 'strings', 'file'
    )
    def test_file_is_relocatable(source_file, is_relocatable):
        compiler = spack.util.executable.Executable('/usr/bin/gcc')
        executable = str(source_file).replace('.c', '.x')
        compiler_env = {
            'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
        }
        compiler(str(source_file), '-o', executable, env=compiler_env)
    
        assert spack.relocate.is_binary(executable)
>       assert spack.relocate.file_is_relocatable(executable) is is_relocatable

relocate.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../relocate.py:678: in file_is_relocatable
    rpaths = patchelf('--print-rpath', file, output=str).strip()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <exe: ['None']>, args = ('--print-rpath', '/tmp/pytest-of-culpo/pytest-43/test_file_is_relocatable_False0/non_relocatable.x'), kwargs = {}
env_arg = None
env = {'BASH_FUNC_spack%%': '() {  if [ -n "${ZSH_VERSION:-}" ]; then\n emulate -L sh;\n fi;\n _sp_flags="";\n while [ ! -z ... 'CLUTTER_IM_MODULE': 'xim', 'COLORTERM': 'truecolor', 'DBUS_SESSION_BUS_ADDRESS': 'unix:path=/run/user/1000/bus', ...}
fail_on_error = True, ignore_errors = (), input = None, output = <type 'str'>, error = None, streamify = <function streamify at 0x7fd09ef7bcf8>
ostream = -1, close_ostream = False

    def __call__(self, *args, **kwargs):
        """Run this executable in a subprocess.
    
            Parameters:
                *args (str): Command-line arguments to the executable to run
    
            Keyword Arguments:
                _dump_env (dict): Dict to be set to the environment actually
                    used (envisaged for testing purposes only)
                env (dict): The environment to run the executable with
                extra_env (dict): Extra items to add to the environment
                    (neither requires nor precludes env)
                fail_on_error (bool): Raise an exception if the subprocess returns
                    an error. Default is True. The return code is available as
                    ``exe.returncode``
                ignore_errors (int or list): A list of error codes to ignore.
                    If these codes are returned, this process will not raise
                    an exception even if ``fail_on_error`` is set to ``True``
                input: Where to read stdin from
                output: Where to send stdout
                error: Where to send stderr
    
            Accepted values for input, output, and error:
    
            * python streams, e.g. open Python file objects, or ``os.devnull``
            * filenames, which will be automatically opened for writing
            * ``str``, as in the Python string type. If you set these to ``str``,
              output and error will be written to pipes and returned as a string.
              If both ``output`` and ``error`` are set to ``str``, then one string
              is returned containing output concatenated with error. Not valid
              for ``input``
    
            By default, the subprocess inherits the parent's file descriptors.
    
            """
        # Environment
        env_arg = kwargs.get('env', None)
        if env_arg is None:
            env = os.environ.copy()
            env.update(self.default_env)
        else:
            env = self.default_env.copy()
            env.update(env_arg)
        env.update(kwargs.get('extra_env', {}))
        if '_dump_env' in kwargs:
            kwargs['_dump_env'].clear()
            kwargs['_dump_env'].update(env)
    
        fail_on_error = kwargs.pop('fail_on_error', True)
        ignore_errors = kwargs.pop('ignore_errors', ())
    
        # If they just want to ignore one error code, make it a tuple.
        if isinstance(ignore_errors, int):
            ignore_errors = (ignore_errors, )
    
        input  = kwargs.pop('input',  None)
        output = kwargs.pop('output', None)
        error  = kwargs.pop('error',  None)
    
        if input is str:
            raise ValueError('Cannot use `str` as input stream.')
    
        def streamify(arg, mode):
            if isinstance(arg, string_types):
                return open(arg, mode), True
            elif arg is str:
                return subprocess.PIPE, False
            else:
                return arg, False
    
        ostream, close_ostream = streamify(output, 'w')
        estream, close_estream = streamify(error,  'w')
        istream, close_istream = streamify(input,  'r')
    
        quoted_args = [arg for arg in args if re.search(r'^"|^\'|"$|\'$', arg)]
        if quoted_args:
            tty.warn(
                "Quotes in command arguments can confuse scripts like"
                " configure.",
                "The following arguments may cause problems when executed:",
                str("\n".join(["    " + arg for arg in quoted_args])),
                "Quotes aren't needed because spack doesn't use a shell.",
                "Consider removing them")
    
        cmd = self.exe + list(args)
    
        cmd_line = "'%s'" % "' '".join(
            map(lambda arg: arg.replace("'", "'\"'\"'"), cmd))
    
        tty.debug(cmd_line)
    
        try:
            proc = subprocess.Popen(
                cmd,
                stdin=istream,
                stderr=estream,
                stdout=ostream,
                env=env)
            out, err = proc.communicate()
    
            result = None
            if output is str or error is str:
                result = ''
                if output is str:
                    result += text_type(out.decode('utf-8'))
                if error is str:
                    result += text_type(err.decode('utf-8'))
    
            rc = self.returncode = proc.returncode
            if fail_on_error and rc != 0 and (rc not in ignore_errors):
                long_msg = cmd_line
                if result:
                    # If the output is not captured in the result, it will have
                    # been stored either in the specified files (e.g. if
                    # 'output' specifies a file) or written to the parent's
                    # stdout/stderr (e.g. if 'output' is not specified)
                    long_msg += '\n' + result
    
                raise ProcessError('Command exited with status %d:' %
                                   proc.returncode, long_msg)
    
            return result
    
        except OSError as e:
            raise ProcessError(
>               '%s: %s' % (self.exe[0], e.strerror), 'Command: ' + cmd_line)
E           ProcessError: None: No such file or directory
E               Command: 'None' '--print-rpath' '/tmp/pytest-of-culpo/pytest-43/test_file_is_relocatable_False0/non_relocatable.x'

../util/executable.py:195: ProcessError
============================================================= 1800 tests deselected ==============================================================
========================================= 2 failed, 3 passed, 1 skipped, 1800 deselected in 4.20 seconds =========================================
```

</details>

### Information on your system

EDIT: **I have `patchelf` installed on the system**. Running `git bisect` points to a940ff34d745e0830f9e40305698bfd85d65fbd3:
```console
a940ff34d745e0830f9e40305698bfd85d65fbd3 is the first bad commit
commit a940ff34d745e0830f9e40305698bfd85d65fbd3
Author: Patrick Gartung <gartung@fnal.gov>
Date:   Thu Sep 19 17:16:26 2019 -0500

    Put back the use of otool and install_name_tool when running on macOS. Only use machotools on linux. (#12867)
    
    Move verbose messages to debug level
    
    get_patchelf should return None for test platform as well because create_buildinfo invokes patchelf to get rpaths.

:040000 040000 93e9e70d3d3747e508d2f124429bd925951ce755 dae8128bce5520a4e80ec9704dcb9ec249d38b34 M      lib
``` 
If it matters:
```console
$ uname -a
Linux nuvolari 5.0.0-29-generic #31~18.04.1-Ubuntu SMP Thu Sep 12 18:29:21 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by alalazo** [comment]
Installing `patchelf` on the system doesn't make a difference.

**Comment by gartung** [comment]
Correct I do not have patchelf in PATH. I changed get_patchelf() to return None if the arch is test because it was trying to install patchelf for tests.

**Comment by alalazo** [comment]
> Correct I do not have patchelf in PATH. I changed get_patchelf() to return None if the arch is test because it was trying to install patchelf for tests

Which test is trying to do that? I added a decorator to avoid that issue:
```python
@pytest.mark.requires_executables(
    '/usr/bin/gcc', 'patchelf', 'strings', 'file'
)
def test_file_is_relocatable(source_file, is_relocatable):
    ...
```
and that's what is triggering the error when `patchelf` is found (because now `get_patchelf` will return `None`). Was the test not protected by this decorator or is there some other issue?

**Comment by gartung** [comment]
I had added get_patchelf() to lib/spack/spack/binary_distribution.py and I needed it to skip during tests. I reverted that change before merging. Perhaps i should also remove the return of None for tests.  Can you test if commenting out or removing this line fixes your case:
https://github.com/spack/spack/blob/a940ff34d745e0830f9e40305698bfd85d65fbd3/lib/spack/spack/relocate.py#L69

**Comment by gartung** [comment]
I should have added the decorator you mentioned to the test instead.

**Comment by alalazo** [comment]
Fixed by #12925
